### ISSUE TITLE
SSH with default password enabled by default.

### ISSUE THREAD

**Issue by vext01** [osmc/osmc#213]
Hi,

Installed OSMC on a raspberry pi2 last week. Impressed, but found that sshd was enabled by default and with a default password (see https://osmc.tv/help/wiki/login). This is a security issue. I would recommend:

 * disabling sshd by default.
 * forcing the user to set a password on first boot.
 * both of the above.

I installed via raw image. I don't know if one of the other installation types including a password reset.

Thanks

**Comment by samnazarko** [comment]
* OSMC's SSH service can be disabled via My OSMC -> Services. It is on by default because by the time a user needs to use it, they may have their system in a state that prevents them getting CLI access.
* Forcing the user to set a password will create two issues: they may do it without realising they need to recall it later; and we do not assume that a user has access to a keyboard input device on first boot.

OSMC is designed to be behind NAT by default. If you wish to make it world facing (for example, by exposing port 22), then you should indeed change the default password.

We believe that as OSMC is not publicly facing and it is possible to easily and disable the SSH service on demand sufficient care has been undertaken. Further efforts to secure OSMC should be conducted by the user.

**Comment by vext01** [comment]
I understand your motivation, but I disagree.

> Further efforts to secure OSMC should be conducted by the user.

> We do not assume that a user has access to a keyboard input device on first boot.

Given your audience, I would consider SSH access as an "advanced" feature that most users will not need or even know what it is.  If you can't ask the user to change their password -- I take your point on this -- you should disable sshd by default. Then advanced users can plug in a USB keyboard to set up a password (via a console keyboard shortcut or a menu item or something).

> OSMC is designed to be behind NAT by default.

Even if the machine is on a private network, it is not hard to imagine situations where visitors in the know could exploit an OSMC machine and do nasty things. E.g. disgruntled employees, mischievous students, etc.

Thanks

**Comment by zabuldon** [comment]
I agree with Edd.
I think best way is a create small configurator, like in openelec, and
provide choice to user at first run

**Comment by samnazarko** [comment]
>I think best way is a create small configurator, like in openelec, and

This can already be changed via My OSMC

>Even if the machine is on a private network, it is not hard to imagine situations where visitors in the know could exploit an OSMC machine and do nasty things. E.g. disgruntled employees, mischievous students, etc.

If you have access to the private network then compromising the OSMC system is arguably not the biggest issue, unless it in turn has access to systems that you are trying to keep secure.

If we disable SSH, then where should the line be drawn? There is a libmicrohttpd service on Port 80 running with a JSON-RPC interface. Unfortunately, there is a Kodi 0-day vulnerability which allows a system to be compromised with this. 

There is a difficult compromise to make between achieving security and ease of use. In a controlled, and home environment, the latter should be the main focus. 

While SSH is usually an advanced application, we often need logs from users or need them to run commands and most of these users are not well versed with the CLI

**Comment by vext01** [comment]
> If you have access to the private network then compromising the OSMC system is arguably not the biggest issue, unless it in turn has access to systems that you are trying to keep secure.

One of OSMC's primary functions is to act as a network client. It deals with with samba credentials, youtube passwords etc. You don't want anyone with access to your network with access to these credentials.

> If we disable SSH, then where should the line be drawn? There is a libmicrohttpd service on Port 80 running with a JSON-RPC interface.

Unless that is essential, I would suggest that this should be off by default too.

> Unfortunately, there is a Kodi 0-day vulnerability which allows a system to be compromised with this.

That is unfortunate and I hope Kodi upstream are working to fix that. But just because there is one vulnerability it doesn't mean it's OK to give up on basic security.

> While SSH is usually an advanced application, we often need logs from users or need them to run commands and most of these users are not well versed with the CLI.

I appreciate this, but it doesn't mean you should compromise their security by default.

I was thinking more about this, and specifically the no-keyboard situation. OSMC has a on-screen keyboard that could be used to set the password on first boot. It only has to be done once. Alternatively you could just start with ssh off and only ask them to set a password upon it's first enablement. Once the password is set, users could be offered an option to change it later from a settings menu, thus preventing them from locking themselves out should they forget it and need to use ssh later to access logs etc.

Sorry to be so persistent, but I think this is important. I won't argue further, but I hope you reconsider.

Overall OSMC is a great OS for a media center and I thank you for your efforts.

**Comment by samnazarko** [comment]
>Once the password is set, users could be offered an option to change it later from a settings menu, thus preventing them from locking themselves out 

I'm not sure on this. As it stands, OSMC is insecure in the regard that there is a default password and SSH is running out of the box. However permitting users to change the OSMC password without needing the original is not a good idea. Anyone with access to the device can then still cause a problem. As it stands, we don't allow passwords to be changed unless the user knows the original. 

I would accept a PR to limit SSH connectivity to the subnet only, and I feel that this addresses the most dangerous potential exploit: being attacked from the outside world when port forwarded.

**Comment by samnazarko** [comment]
>Unless that is essential, I would suggest that this should be off by default too.

The web server is on so that people with a remote or keyboard are able to control Kodi out of the box. Otherwise they have no way of controlling the device at all.

**Comment by vext01** [comment]
> I'm not sure on this. As it stands, OSMC is insecure in the regard that there is a default password and SSH is running out of the box. However permitting users to change the OSMC password without needing the original is not a good idea.

Well, as it stands on a new install attackers *always* know the original password, as it is a fixed default,  thus they can log in over SSH and change it via `passwd`. This will of course ask for the current password, but they already know it. At-least the method I proposed only allows the password to be changed if the user is physically holding the remote.

**Comment by samnazarko** [comment]
I am not convinced that substituting one potential security risk for another is a good idea. There is still the possibility to change the password using the HTTP API (as an attacker could input it via a remote control).

Happy to accept PR for restricting SSH to local connection only and I think this will mitigate things well.

**Comment by KODeKarnage** [comment]
I don't think that changing the password in Settings would work. How exactly would you restrict the password reset to only be valid from the settings menu? It would have to call some external system functionality anyway, and all that would be visible in the code. And what about the users that can't access the settings menu? What about the users who are having problems interfacing? You would be forcing them to pave and reinstall for little to no security gain.

It is incorrect to assume that SSH is an "advanced users" tool, when even novice users can use it to solve their issues. I don't think you have sufficient appreciation of the time and effort that having a default password has saved users.

The default password is only a problem if you have the device open to the net or otherwise have your network compromised. In the first case, that would only happen if the user chooses to do it, and if they do that then they can take responsibility for their own security. In the second case, having a default password on your media player is the very least of your problems.

In effect, your complaint is that the lock on the jewelry box is flimsy and would pose no impediment to the anyone who manages to break into the locked safe where it is kept.

**Comment by vext01** [comment]
I think you are making too many assumptions about the kinds of network your users -- many of whom will be novices -- will install OSMC into. The truth is, we have no idea which kinds of networks users will deploy into.

Just because a network is a LAN it doesn't mean it should be trusted. For example, can you not imagine students installing OSMC onto a dormitory network shared by thousands of other (unknown) students?

I'm not going to argue further, but I maintain that anything connected to a network should be locked down by default.

Best of luck with OSMC. Cheers.

**Comment by malkavi** [comment]
Then users need to learn how to use it, reading the wiki. When I fresh install OSMC into a rasp I usually use ssh to copy some files, like lircd.conf. I don't want to use a usb keyboard or any other hardware. We can assume that the 99% of the users are behind a router. And the others can change their password.

**Comment by davmac314** [comment]
You leave a service like SSH running with a default password and I pretty much guarantee you just created a botnet. Well done. News flash: your users won't do what you want or expect them to do. A proportion of the devices will get exposed to the wider world, through accident or negligence - yes, even a proportion of that alleged 99% that sit behind a router; you can't prevent this, but you can mitigate the damage, and it would be irresponsible not to do so.

The fix isn't even that hard: enable the HTTP interface but insist that the password be changed the first time it is accessed, or (if it happens beforehand) the first time the on-screen UI is used with a keyboard/remote. (Ideally, disable SSH until the password is changed, but if you oblige the user to change the password then even this isn't so necessary).

Your flippant comment that "users need to learn how to use it, reading the wiki" is exactly why asking them to set a password that they will remember should not be a problem. By leaving this unresolved you have created headaches for network administrators everywhere just to make things a smidgin more convenient for your users.

**Comment by malkavi** [comment]
I still think that OSMC is an OS for "home", a place without net admin. But a "solution" can be to display a message, at the first boot, with a link to the wiki, something like "how to secure your OSMC". 

But you can't disable ssh by default, ssh is the common way to access the rasp. Also can be added a warning to the ssh interface.

Anyway, maybe https://discourse.osmc.tv is a better place to talk about this and know what other users think.

**Comment by davmac314** [comment]
> But a "solution" can be to display a message, at the first boot, with a link to the wiki, something like "how to secure your OSMC". 

Yeah, or you could just get them to set a password and at that stage it _is_ secure.

> But you can't disable ssh by default, ssh is the common way to access the rasp.

That's no logical connection between the two parts of that sentence. I agree with your premise that ssh is a common way to access a Pi (though I don't think that's relevant) but I disagree with your conclusion. Of course you can disable ssh by default.

> Also can be added a warning to the ssh interface.

... which would be of no use to anyone not using ssh themselves. Maybe you should instead have a message to would-be hackers asking them to "please go away"? It seems like it would be just as effective :stuck_out_tongue:

(I'm done with the discussion. I hope you guys see the light eventually. Good luck).

**Comment by DBMandrake** [comment]
>Yeah, or you could just get them to set a password and at that stage it is secure.

So the same users you don't trust to disable SSH or pro-actively change the password themselves, you trust to choose a strong password that isn't their pets name, and enter it laboriously using an on-screen keyboard and remote control, and then not forget the password later for when they actually need it to rescue their system when they screw it up ? Riiiiiiiight... :stuck_out_tongue_winking_eye:

**Comment by davmac314** [comment]
> So the same users you don't trust to disable SSH or pro-actively change the password themselves, you trust to choose a strong password that isn't their pets name, and enter it laboriously using an on-screen keyboard and remote control, and then not forget the password later for when they actually need it to rescue their system when they screw it up ?

I trust them to do none of those things, but that doesn't matter; all I need them to do is change it to something that a kid with a script and a IP port scanner can't trivially get into the system with - I.e. something that's not the default password.

(I didn't want to come back to this discussion, but I have to admit I'm easily goaded).

I mean, sure, you can keep this issue closed and continue to fly in face of long-established security practice and wisdom, or you can make users choose a password which they might forget at the possible cost of having to re-image their device. If that really matters, then let me assure you, at least they won't forget it the second time.

**Comment by samnazarko** [comment]
>I mean, sure, you can keep this issue closed and continue to fly in face of long-established security practice and wisdom

You seem concerned about access from an external network. As said twice now, I will accept a PR to change SSH behaviour to only accept connections on the same LAN segment

Sam

**Comment by davmac314** [comment]
> You seem concerned about access from an external network. As said twice now, I will accept a PR to change SSH behaviour to only accept connections on the same LAN segment

As I said earlier:

> A proportion of the devices will get exposed to the wider world, through accident or negligence - yes, even a proportion of that alleged 99% that sit behind a router; you can't prevent this, but you can mitigate the damage, and it would be irresponsible not to do so.

Limiting SSH connections to the same subnet might help, but I don't see why you should half-fix the problem when there is a full solution.

**Comment by KODeKarnage** [comment]
There is no such thing as 100% secure, at some point you have to be ok with good enough, and blithely ignoring the costs of "mitigation" doesn't make disappear.

**Comment by davmac314** [comment]
> There is no such thing as 100% secure, at some point you have to be ok with good enough

I don't believe that having a default password _is_ good enough. In the scale of things, it's barely better than having no password at all.

> and blithely ignoring the costs of "mitigation" doesn't make them disappear

What, the "cost" of users needing to actually remember their password? In my view cost/benefit analysis weighs heavily in favor of not having a default password. I'm surprised that anyone would disagree with this, but there it is, I guess.

**Comment by KODeKarnage** [comment]
Relying on the password alone isnt good enough. Relying on your network security IS good enough.

And the cost isn't people REMEMBERING their password. The cost is people FORGETTING their password.

Both of those facts are obvious to anyone who has actually considered the issue. Maybe come back when you have given this more than a few seconds thought.

**Comment by davmac314** [comment]
> Relying on the password alone isnt good enough. Relying on your network security IS good enough.

I didn't say that relying on the password alone is good enough. OTOH I disagree that relying on "your" network security is good enough, partly because I don't know who "you" (the users) are and partly because there can be holes in network security due to accident, negligence, or ignorance.

> And the cost isn't people REMEMBERING their password. The cost is people FORGETTING their password.

You're just nit-picking and playing with words. The cost to the user is that they are obliged to remember the password or suffer the consequence of forgetting.

 > Both of those facts are obvious to anyone who has actually considered the issue. Maybe come back when you have given this more than a few seconds thought.

I don't see the point in arguing with you at this stage; you're being ignorant and rude.

**Comment by KODeKarnage** [comment]
No great loss.

**Comment by joepie91** [comment]
> There is no such thing as 100% secure, at some point you have to be ok with good enough, and blithely ignoring the costs of "mitigation" doesn't make them disappear.

And it is well understood within the security community that the only correct solution is secure-by-default, even if that is not the right configuration for 100% of cases.

I'm honestly not even sure why there's even still a discussion here. Exposing an SSHd with default credentials to the network is unacceptable, end of story.

**Comment by samnazarko** [comment]
Little late to the discussion here.

OSMC has already implemented an option to enable or disable SSH on first run.

**Comment by joepie91** [comment]
Yes, I just tried out OSMC on my new RPi3 - and I don't recall seeing a clear explanation of the risks that it poses, nor does it ask the user to configure a password (or even give them a random string of numbers/letters to write down!). This doesn't really fully solve the issue.

I can't really easily go back and see whether it makes any mention of it *at all*, but it's definitely not explained clearly enough.

**Comment by samnazarko** [comment]
>or even give them a random string of numbers/letters to write down

We'd also then need to suggest that writing down a password should be avoided, and if this must be done it should be stored in a secure place.

We've changed the default behaviour of enabling SSH without informing the user and now make things easer than ever to adjust on first boot.

However, outlandish statements such as 

>pretty much guarantee you just created a botnet 

make it increasingly challenging for us to take further comments on this issue seriously. 

Save yourself, and myself time. I'm not going to change the default behaviour. If you'd like to bring the concerns of running SSH to light then I'd encourage a PR to the walkthrough to add a small background this.

**Comment by securitymad84** [comment]
Congratulations! You created a bot net you just don't know it yet

You also won the award for most insecure piece of software 2016 at our conference

**Comment by joepie91** [comment]
> We'd also then need to suggest that writing down a password should be avoided, and if this must be done it should be stored in a secure place.

Not really, just tell them to stick the password to the bottom of the Pi, or something. The threat model for the average user is preventing network-based attacks at home - the moment there's physical access to the device, you've basically already lost anyway.

> However, outlandish statements such as "pretty much guarantee you just created a botnet" make it increasingly challenging for us to take further comments on this issue seriously.

Those statements are not outlandish, if a bit brash. That *is* what happens when you ship a device or OS with default SSH credentials.

> If you'd like to bring the concerns of running SSH to light then I'd encourage a PR to the walkthrough to add a small background this.

Where in the repository do I find the relevant code?

**Comment by KODeKarnage** [comment]
> You also won the award for most insecure piece of software 2016 at our conference.

You mean that conference being run by someone with a brand new github account? Yeah, that'll be huge.

**Comment by samnazarko** [comment]
>The threat model for the average user is preventing network-based attacks at home

Actually, the threat model for the average user is outside connection to the device. This will likely only occur due to incorrect port-forwarding, or a DHCP lease expiring and a port forward for 22 'leaking' from one intended device to another.

>Those statements are not outlandish

They absolutely are. Are you stating that we have created a botnet, and if so, do you have any evidence to support this? It's quite a statement to make. 

>Where in the repository do I find the relevant code?

See package/mediacenter-addon-osmc.

**Comment by securitymad84** [comment]
I take security very seriously and made this account so my real identity would not be known

**Comment by KODeKarnage** [comment]
> I take security very seriously and made this account so my real identity would not be known

Yeah, not because you are a pathetic loser or anything.

Go back to organising that "conference" and say hi to that "girlfriend from Canada" for me.

**Comment by joepie91** [comment]
> Actually, the threat model for the average user is outside connection to the device. This will likely only occur due to incorrect port-forwarding, or a DHCP lease expiring and a port forward for 22 'leaking' from one intended device to another.

Absolutely not. IPv6 adoption is rapidly growing, and will by-default expose many devices on many networks to the public internet. *NAT is not a firewall* - don't be like Canon, whose PIXMA printers happily allowed anybody to upgrade the firmware, and which also happily announced themselves on IPv6 addresses on many networks; mine included.

Not to mention the people who *do* actually get dedicated IPv4 addresses for each device.

> They absolutely are. Are you stating that we have created a botnet, and if so, do you have any evidence to support this? It's quite a statement to make.

I'm not sure whether this is intentional or not, but you are reading that statement *far* too literally. When it's said that "you created a botnet", what is meant is that it's only a matter of time until somebody scans the internet for OSMC devices and automatically compromises every device found.

Indirectly, your decision to use default SSH credentials would have made it possible to do so in the first place. This is the exact same problem that many other embedded hardware vendors have had in the past (although they sometimes used telnet instead), and has played a *significant* role in DDoS attacks throughout the past decade or so.

> See package/mediacenter-addon-osmc.

I'll probably have a look somewhere next week.

**Comment by DBMandrake** [comment]
>NAT is not a firewall 

Incorrect. Network Address Translation is inherently a stateful firewall. A firewall that by default is completely permissive in the *outgoing* direction, granted, but a stateful firewall nevertheless.

Unsolicited inbound connections cannot pass through NAT (assuming no static port forwarding has been set up) because no state exists to map those unsolicited connections back to internal hosts. It might be trendy to say NAT is not a firewall, but it's simply not true.

> Absolutely not. IPv6 adoption is rapidly growing, and will by-default expose many devices on many networks to the public internet.

And exactly what do you base this wild assumption on ? Do you seriously believe that the world will move from having most devices being relatively safe behind NAT firewalls to every individual device not only having a public IP but border routers providing no firewalling or filtering of traffic at all ? Really ?

Just because IPv6 makes full unfettered bidirectional connectivity theoretically possible does not mean it is a good idea, nor does it mean it will happen this way. It is perfectly practical for a border router to provide stateful firewalling without performing Network Address Translation. You gain all the benefits of blocking unsolicited inbound connections whilst still freely allowing outbound connections, but without being stuck with one IP.

Network Address Translation requires a stateful firewall to function, stateful firewalls do not require Network Address Translation.

Wake me up when IPv6 finally arrives for real - I've been preparing for IPv6 for nearly 20 years now and I'm still waiting...

**Comment by joepie91** [comment]
> Incorrect. Network Address Translation is inherently a stateful firewall. A firewall that by default is completely permissive in the outgoing direction, granted, but a stateful firewall nevertheless.

This is a *side-effect*, not its *purpose*. Which is a critical difference, because that means you cannot rely on it for security purposes. Rule one in security may be not to roll your own, but rule two is not to use things for purposes other than what they were intended for, because there's going to be a mismatch in assumptions at some point.

As for IPv6: __I have literally just told you that this is precisely what happened on my network.__ By default. With ISP-supplied hardware. On a residential connection.

Please take your "well then they just have to configure it correctly" nonsense elsewhere. No, consumer routers often do not ship with a firewall by default, and no matter how hard you shout that "well, you can configure that", that *is not going to change the reality that most people will not use it*.

**Comment by SN4T14** [comment]
>Unsolicited inbound connections cannot pass through NAT (assuming no static port forwarding has been set up) because no state exists to map those unsolicited connections back to internal hosts.

Have you heard of [TCP hole punching](https://en.wikipedia.org/wiki/TCP_hole_punching)?

>And exactly what do you base this wild assumption on ? Do you seriously believe that the world will move from having most devices being relatively safe behind NAT firewalls to every individual device not only having a public IP but border routers providing no firewalling or filtering of traffic at all ? Really ?

Yes, this is precisely what is already happening with IPv6, pretty much all consumer routers don't do NAT for IPv6. And contrary to what you seem to be saying, almost all OSes ship with a sane-ish firewall by default. This means that it's completely sane to disable NAT when possible, especially since you no longer violate the [end-to-end principle](https://en.wikipedia.org/wiki/End-to-end_principle). Of course, not having to NAT every packet means you don't need nearly as fast a CPU on the router, and we all know how router manufacturers love cutting costs in the hardware department.

>It is perfectly practical for a border router to provide stateful firewalling without performing Network Address Translation.

So is running a normal firewall on your device... Oh wait, we *already do that*.

>Network Address Translation requires a stateful firewall to function, stateful firewalls do not require Network Address Translation.

Wait, but I thought you said this?

>Network Address Translation is inherently a stateful firewall.

Looks like you're contradicting yourself there, either NAT *is* a stateful firewall, or it *requires one* to function. Either you're claiming NAT *is* a firewall, and it's not according to the definition, so you shouldn't rely on it for security, it's an edge-case that's not guaranteed, and may not even be there if you're NATing. Or you're claiming NAT *requires* a stateful firewall, so let's say it did, then router manufacturers won't ever bother with NAT for IPv6, and will do away with the stateful firewall as well because devices already ship with sane firewalls anyway.

>Wake me up when IPv6 finally arrives for real - I've been preparing for IPv6 for nearly 20 years now and I'm still waiting...

The US is getting close to [25% adoption](https://www.google.com/intl/en/ipv6/statistics.html#tab=per-country-ipv6-adoption&tab=per-country-ipv6-adoption), and if global adoption rates continue the current trend, those will be around 20% by the end of the year as well. So yes, IPv6 *has* arrived for real, and by the end of this year, *at least* 20% of your users will be vulnerable, stop trying to deflect blame and **fix this**.

**Comment by KODeKarnage** [comment]
Does the SSH server listen on ipv6?

Because if it doesn't, then would that make the last couple of posts entirely pointless?

**Comment by vext01** [comment]
Hi,

OP here. I'm surprised to see this thread still going.

There's a lot of argumentation in this thread which detracts from the issue: enabling sshd by default, with a default password, is a trivial attack surface to exploit.

Make no assumptions about the kind network your users will deploy into. I would expect OSMC installs to be secured on the *local* network by default. Any additional network reachability only increases the chances that the existing attack surface will be exploited. (Yes, internet reachability increases the threat of attack. No, NAT is not security. Yes, IPV6 is widely used. However these are all secondary issues).

FWIW, I spoke with Sam via private email, and we had agreed that there should be a setup screen which asks the user if sshd should be enabled (hopefully defaulting to off, and with a big warning message for inexperienced users). I tried a fresh OSMC install a month or two ago and didn't see any such option during install. I did manage to find some changes in git which perhaps suggested the OSMC team were preparing to change the default SSH configuration, but I'm not sure if that code was complete.

@samnazarko can you give us an update?

Cheers

**Comment by samnazarko** [comment]
Since around November, OSMC will prompt the user whether to enable SSH or not on first run. Did you use a recent disk image, or did you dd an old one and the update manually?

Sent from a mobile device, so please excuse the brevity of this reply

On 13 Mar 2016, at 09:09, Edd Barrett <notifications@github.com<mailto:notifications@github.com>> wrote:


Hi,

OP here. I'm surprised to see this thread still going.

There's a lot of argumentation in this thread which detracts from the issue: enabling sshd by default, with a default password, is a trivial attack surface to exploit.

Make no assumptions about the kind network your users will deploy into. I would expect OSMC installs to be secured on the local network by default. Any additional network reachability only increases the chances that the existing attack surface will be exploited. (Yes, internet reachability increases the threat of attack. No, NAT is not security. Yes, IPV6 is widely used. However these are all secondary issues).

FWIW, I spoke with Sam via private email, and we had agreed that there should be a setup screen which asks the user if sshd should be enabled (hopefully defaulting to off, and with a big warning message for inexperienced users). I tried a fresh OSMC install a month or two ago and didn't see any such option during install. I did manage to find some changes in git which perhaps suggested the OSMC team were preparing to change the default SSH configuration, but I'm not sure if that code was complete.

@samnazarko<https://github.com/samnazarko> can you give us an update?

Cheers

-

**Comment by DBMandrake** [comment]
>Have you heard of TCP hole punching?

Yes of course I have. Have you looked up the dictionary definition of "unsolicited" ?

TCP hole punching does absolutely *nothing* to reinforce your argument, because it requires *active participation by both ends to establish a connection*. How about reading the article you link to and actually think through whether it is relevant to the discussion before trying to use it to prop up your argument.

NAT traversal techniques like this have been around a long time (especially UDP based ones used in gaming) and are a technique to allow *two willing and participating* endpoints to establish a direct connection to each other when they are both behind NAT. (Temporarily using a 3rd party not behind NAT to help figure out what port numbers to use)

Now lets consider the case of an SSH server minding its own business listening on Port 22 behind a NAT firewall that is not port forwarding to it. How exactly is an outside attacker going to use this to gain access to the SSH server ? Answer - they can't.

To use TCP hole punching the SSH server would have to be rewritten to be a complicit and willing partner in the hole punching process - not only would it need to send outbound SYN packets, it would also need to be connecting to a pre-defined 3rd party rendezvous to figure out what port to send its SYN's to. Clearly the SSH server in OSMC does not do this so your point is completely irrelevant.

>Yes, this is precisely what is already happening with IPv6, pretty much all consumer routers don't do NAT for IPv6.

Where did I say that routers should be doing NAT for IPv6 ? I said they should continue to provide stateful firewalling where they block incoming unsolicited connections by default. I was pretty clear about this. Perhaps you don't understand the distinction between a stateful firewall and NAT ?

>And contrary to what you seem to be saying, almost all OSes ship with a sane-ish firewall by default.

Uh huh. You mean like all the IOT devices that run web severs with default (or no) password ? It's not so much main stream operating systems that you need to worry about, but devices like these.

>This means that it's completely sane to disable NAT when possible, especially since you no longer violate the end-to-end principle. Of course, not having to NAT every packet means you don't need nearly as fast a CPU on the router, and we all know how router manufacturers love cutting costs in the hardware department.

Again, you confuse NAT with stateful firewalling. You also make the false assumption that NAT is a high overhead process - it is not. It is performed in the IP routing layer of the kernel on a router and basically only requires a lookup stable to store state, and during packet routing the source port and IP address are rewritten. If you are only doing stateful firewalling the process is even simpler - now you only need to keep a state table and no source IP/port rewriting needs to be done.

>>Network Address Translation requires a stateful firewall to function, stateful firewalls do not require Network Address Translation.

>Wait, but I thought you said this?

>>Network Address Translation is inherently a stateful firewall.

>Looks like you're contradicting yourself there, either NAT is a stateful firewall, or it requires one to function. Either you're claiming NAT is a firewall, and it's not according to the definition, so you shouldn't rely on it for security, it's an edge-case that's not guaranteed, and may not even be there if you're NATing.

I'm not contradicting myself at all - all you're doing here is showing your ignorance of how NAT and stateful firewalls work. A stateful firewall is something where decisions on whether to allow or block packets is based on state that is created and saved in a table in response to previously sent packets, usually in the other direction.

For example when you make an outgoing TCP connection from a LAN client to a server on the Internet as well as (by default) allowing that outgoing packet through, the source and destination IP and ports are stored in a lookup table so that when the reply comes back the packet is allowed to enter. If the exact same incoming packet had been received without that initial outgoing packet to generate state, it would have been ignored or rejected. That is a basic stateful firewall.

NAT does all this *and* does one extra thing - it also rewrites the source IP address and port of every outgoing packet and stores the translation in the state table. Then when incoming packets return not only can it look up the state table to see whether the packet should be allowed through or dropped, it can then also look up the translation entries to figure out how to rewrite the destination IP and port before routing the packet onto the LAN.

So NAT is a superset of a stateful firewall. It is *not possible* to have a NAT firewall that isn't a stateful firewall simply due to the way it works - if state is not saved it couldn't figure out which client on the LAN to forward packets to or what ports to send to. So your claim "Either you're claiming NAT is a firewall, and it's not according to the definition, so you shouldn't rely on it for security, it's an edge-case that's not guaranteed" is completely false.

>The US is getting close to 25% adoption, and if global adoption rates continue the current trend, those will be around 20% by the end of the year as well. So yes, IPv6 has arrived for real, and by the end of this year, at least 20% of your users will be vulnerable, stop trying to deflect blame and fix this.

Fix what exactly ? Clearly you are not an OSMC user and have not done any actual testing or you would have discovered that we **disable** IPv6 on OSMC default...so your fears of exposed IPv6 SSH are completely without factual basis.

https://github.com/osmc/osmc/blob/master/package/connman-osmc/patches/all-004-disable-ipv6-by-default.patch

This patch disables IPv6 by default on all new interfaces that connman detects/configures and was added **9 months ago**.

Furthermore:

https://github.com/osmc/osmc/blob/master/package/connman-osmc/files/DEBIAN/postinst#L17-L24

This script was added at the same time and retroactively **disabled** all existing IPv6 configurations during the upgrade process. So any OSMC user who has installed *or* updated in the last 9 months does not have IPv6 enabled.

There is no GUI configuration for IPv6 either, so the only way to turn it back on is to use the connmanctl command line tool, which your average non-technical user is not going to do. In the 9 months since we did this I can only remember two forum posts saying "what happened to my IPv6" connection, so either the few power users who wanted to use IPv6 figured out how to re-enable it via the command line on their own, or nobody uses IPv6....

**Comment by vext01** [comment]
@samnazarko Thanks for getting back to me. I used the latest image at the time, which admittedly was a month or so ago. I have some time today and a spare rpi2, so I can install afresh and report back.

Cheers

**Comment by KODeKarnage** [comment]
I tried if earlier myself. The ssh panel DID display.

**Comment by vext01** [comment]
OK, I can confirm that OSMC now has a SSH configuration dialog on first run.

In it's initial state, it looks like this:

![2](https://cloud.githubusercontent.com/assets/604955/13728718/0d2dd2ea-e919-11e5-8a17-37bbd3ab6771.jpg)

This is a step in the right direction, as now at least users in the know of ssh now can choose to turn it off.

However, I'm sorry to say that I still find the defaults a little dangerous. The box is checked by default (as shown in the picture), and the prose above suggests that having an ssh daemon running is beneficial (there is no mention of the security ramifications or the fact that there is a default password). Inexperienced users are bound to just hit "accept" blindly.

I'm not saying openelec's solution is perfect, but I find it much more palettable:

![1](https://cloud.githubusercontent.com/assets/604955/13728739/c6b47eee-e919-11e5-916c-650d2c7cf2c3.jpg)

 * SSH off by default.
 * Clear indication that it is for advanced users only.
 * Clear inidcation of default password.

If it were me, I would go one step further and have the user input or generate a password if the user opted in to enabling ssh. Regardless, I think the openelec team's approach is a lot less prone to security problems.

**Comment by samnazarko** [comment]
I'm sure we can add some more verbosity around SSH and what it does, but I don't agree that OE is more secure (they provide root access, and the password is not changeable)

Sam

Sent from a mobile device, so please excuse the brevity of this reply

On 13 Mar 2016, at 12:53, Edd Barrett <notifications@github.com<mailto:notifications@github.com>> wrote:


OK, I can confirm that OSMC now has a SSH configuration dialog on first run.

In it's initial state, it looks like this:

[2]<https://cloud.githubusercontent.com/assets/604955/13728718/0d2dd2ea-e919-11e5-8a17-37bbd3ab6771.jpg>

This is a step in the right direction, as now at least users in the know of ssh now can choose to turn it off.

However, I'm sorry to say that I still find the defaults a little dangerous. The box is checked by default (as shown in the picture), and the prose above suggests that having an ssh daemon running is beneficial (there is no mention of the security ramifications or the fact that there is a default password). Inexperienced users are bound to just hit "accept" blindly.

I'm not saying openelec's solution is perfect, but I find it much more palettable:

[1]<https://cloud.githubusercontent.com/assets/604955/13728739/c6b47eee-e919-11e5-916c-650d2c7cf2c3.jpg>

  *   SSH off by default.
  *   Clear indication that it is for advanced users only.
  *   Clear inidcation of default password.

If it were me, I would go one step further and have the user input or generate a password if the user opted in to enabling ssh. Regardless, I think the openelec team's approach is a lot less prone to security problems.

-

**Comment by vext01** [comment]
Hey Sam,

More verbosity in the dialog would certainly be welcomed.

I did notice that openelec doesn't allow you to change the password, and I do agree that this is a little odd (why do that?). Regardless, the point is that their *default* settings are secure, and inexperienced users are not encouraged to shoot themselves in the foot.

You could match their default settings by simply having the ssh check box disabled by default, and adding prose to the effect of:

```
Once enabled, the SSH server uses a default password (user=osmc, pass=osmc). SSH should only be enabled by advanced users who understand to risks of such a configuration. If you enable SSH, you are strongly encouraged to change the password.
```

That should do the trick.

Cheers!

**Comment by samnazarko** [comment]
I am sure we can make that change

Sam

Sent from a mobile device, so please excuse the brevity of this reply

On 13 Mar 2016, at 13:32, Edd Barrett <notifications@github.com<mailto:notifications@github.com>> wrote:


Hey Sam,

More verbosity in the dialog would certainly be welcomed.

I did notice that openelec doesn't allow you to change the password, and I do agree that this is a little odd (why do that?). Regardless, the point is that their default settings are secure, and inexperienced users are not encouraged to shoot themselves in the foot.

You could match their default settings by simply having the ssh check box disabled by default, and adding prose to the effect of:

Once enabled, the SSH server uses a default password (user=osmc, pass=osmc). SSH should only be enabled by advanced users who understand to risks of such a configuration. If you enable SSH, you are strongly encouraged to change the password.


That should do the trick.

Cheers!

-

**Comment by davmac314** [comment]
I'm glad that you're addressing this, especially after developer comments such as those above ("Relying on your network security IS good enough", "Maybe come back when you have given this more than a few seconds thought"). It's a shame that so much resistance was put up initially. I hope you can keep OSMC secure, going forward, and also that your community can become less toxic towards users who try to constructively discuss any issues. The irony that the same developer who told me that my opinion was both wrong and worthless has apparently themselves given more than "a few seconds thought" to the issue now (and implemented SSH disable option) has not escaped me.

**Comment by KODeKarnage** [comment]
You said you were surprised anyone would disagree with you, but then spent absolutely no time trying to appreciate why that might be the case. 

("What, the "cost" of users needing to actually remember their password?")

You completely ignored the massive costs of having users effectively locked out of their device and unable to implement fixes. In fact, you dismissed ALL costs as a mere "smidgen".

You either did not spend any time thinking about those costs, or you did spend time but couldn't think of them, or you could think of them but placed zero value in them. THAT is arrogant, and it betrays your feelings toward the people who would pay those costs.

Your fantasies about bot-nets notwithstanding (and going to pay up on that "pretty much guarantee"?) you didn't give this more than a few seconds thought. You embraced your priors and waved away the effects on other people. Your cost-benefit analysis amounted to little more than extolling the benefits and kicking the costs off the scale.

**Comment by KODeKarnage** [comment]
And notice that not once have I ever said that the user being able to change the password would be a bad thing. But there are costs to forcing them to do it, and those should be acknowledged and weighed against the benefits. 

From the very start you treated the costs as non-existent, and insulted the intelligence of people who didn't value them the same.

**Comment by samnazarko** [comment]
I fail to see any value in the above comment. Just an 'I told you so'. Some others have pointed out that we can make things better and even volunteered to PR those changes. You have just sat on the sidelines and fuelled the fire throughout this discussion.

**Comment by leonardder** [comment]
Whatever is being done to increase OSMC security, I strongly disagree to disable SSH by default. As a blind user, there is currently no other way to enable the Kodi Screen Reader other than SSH, unless you have sighted assistance.

**Comment by davmac314** [comment]
> Your cost-benefit analysis amounted to little more than extolling the benefits and kicking the costs off the scale.

And yours completely discounted the cost of having compromised systems.

> From the very start you treated the costs as non-existent,

No, I treated them as insignificant compared to the real cost - of having easily compromisable systems spread throughout the user base - which I still believe they are.

> and insulted the intelligence of people who didn't value them the same.

I call BS on that; it's simply untrue. The record is above for all to see. On the other hand, _you_ insulted _my_ intelligence quite deliberately with your snide "think about it for a few minutes"/"you really haven't thought this through"/"No great loss" remarks.

@samnazarko:

Firstly, yes, that was a bit of an "I told you so", but it was certainly deserved, and at least _I_ refrained from personal insults.

> You have just sat on the sidelines and fuelled the fire throughout this discussion.

Again I call BS. Yep, I used strong statements initially (although those now taking obvious hyperbole designed to stress a particular point as literal statements are surely just as guilty of "fuelling fire", as were comments made by certain others implying my opinions had no merit or were otherwise worthless). Initial comments on this issue showed that your view on security was, in my eyes, lax; Heck, you had already closed this issue, despite what I considered to be a massive security issue having been brought to your attention; I wanted to voice a strong opinion. But at no stage did I resort to personal insults, belittle anyone or their insult their intelligence. Instead, that was done to me - "Maybe come back when you have given this more than a few seconds thought", as if all the legitimate points I had made were nothing but idle fantasies which I had cooked up in "a few seconds". When I said that there was no point continuing an argument with someone who was rude to me, I was told that it was "no great loss". Now you say that _I_ was "fuelling fire"; all I can say is, if you don't want the discussion to get heated, maybe you should step in at the point where personal insults are being hurled.

Frankly, if it helped to get this issue turned around, then I'm not ashamed of making strong statements. This is about security, and it's important, not just for you or me, or even just the users of OSMC. From your statements above it seems like progress is being made, and that's a good thing. I hope that you can understand why I might feel inclined to make strong statements regarding this.

But yes, progress. So, although I feel far from welcome in this community at this stage, let me say thanks, and when it comes to security, please keep it up. It's important, as I'm sure you agree.

**Comment by KODeKarnage** [comment]
> And yours completely discounted the cost of having compromised systems.

Yeah? Then I am sure you will be able to find a quote to back up your point. Go ahead and try.

> No, I treated them as insignificant compared to the real cost - of having easily compromisable systems spread throughout the user base - which I still believe they are.

So those insignificant costs aren't real, according to you. And yet you start the sentence denying you treated the costs as if they weren't real. Not only that, you are trying to frame the BENEFITS as being the "real" costs! NOT having a compromised system is a BENEFIT, genius! The COST cant then be having a compromised system!

And again, you did not give this topic more than a few seconds thought.  You acted from the very start as if the "correct" position should have been obvious to all. You were surprised that anyone would question you... *ahem*, I mean, question it.

Well, things that are obvious don't require more than a few seconds thought, genius.

You got offended because I said directly to your face what you left as a cowardly deniable insinuation towards us.

**Comment by davmac314** [comment]
> Yeah? Then I am sure you will be able to find a quote to back up your point. Go ahead and try.

You want me to find a quote that says "I KodeKarnage am not considering the costs of compromised systems"? Don't be ridiculous. I am able to say that you don't consider that cost because you make no statements showing that you do; the onus to prove otherwise is on you.

> So those insignificant costs aren't real, according to you

Taking every word with a literary dictionary definition just makes this even more painful. "Real cost" as in major, significant, dominant. Yes, those other costs are not dominant.

> NOT having a compromised system is a BENEFIT, genius! The COST cant then be having a compromised system!

A benefit of one alternative is logically the opposite of a cost in another alternative.

> And again, you did not give this topic more than a few seconds thought.

I can only assume that you say this because it allows you to discount my arguments without having to resort to logic and reason, at which point you'd be lost. If that's not the case, then stop making this claim; it's insulting and not constructive.

> You were surprised that anyone would question you... ahem, I mean, question it.

In no way did I suggest that I was surprised someone would question _me in general_. Again, you're resorting to personal attacks instead of reasoned argument, trying to make me look arrogant by twisting my words. Stick to a reasoned discussion.

(The notion that the cost of compromised systems outweighs a certain level of convenience for users? Yes, I was surprised that anyone would question it. Genuinely.)

> You got offended because I said directly to your face what you left as a cowardly deniable insinuation towards us.

What you _construed_ as an insinuation towards you, it seems. If this is at the root of your apparent anger, let me assure you that it was not intended as such. If you had questioned that statement instead of just insulting me, I could have elaborated on this then.

I'm done. I'm sick of your childish attitude and your refusal to address reasoned arguments with suitably reasoned counterarguments.

**Comment by KODeKarnage** [comment]
> I am able to say that you don't consider that cost because you make no statements showing that you do; the onus to prove otherwise is on you.

And I am able to say that you don't consider it wrong to lewdly expose yourself to pine trees because you make no statements showing that you do. Can you see the flaw in your logic NOW?

> A benefit of one alternative is logically the opposite of a cost in another alternative.

Which is a problem when you choose to only consider the benefits of one alternative and the costs of the other. 

> I can only assume that you say this because it allows you to discount my arguments without having to resort to logic and reason, at which point you'd be lost. 

When the problem is YOU discounting the costs to zero, it is a bit telling that you accuse others of this. 

> The notion that the cost of compromised systems outweighs a certain level of convenience for users? Yes, I was surprised that anyone would question it. 

Nobody questioned the notion.  Again, I challenge you to find any quote showing I did. The question was always your definition of "a certain level": ZERO. When shown that your definition of that cost was wrong (the cost is not remembering the password, as you mockingly put it, the cost is related to the far from zero probability that it would be forgotten) you whined about that being semantics! 

Oh but nobody could possibly have a problem with YOUR value-judgements, right? So they must have a problem with the concept of security in its entirety! Those idiots! 

Either it was obvious or it took more than a few seconds thought. You can't have it both ways. 

> I'm done. 

You have been.

**Comment by davmac314** [comment]
> And I am able to say that you don't consider it wrong to lewdly expose yourself to pine trees because you make no statements showing that you do. Can you see the flaw in your logic NOW?

That would be a suitable comparison only if the general theme of the discussion up to that point had been about the morality of doing so. We were discussing cost/benefit of having a default password or not, if you had indeed considered the cost of compromised systems as being of any significance, you had failed to mention it. So much as you claim that I consider user convenience as 0 benefit (which I've never said), I said that you didn't consider the cost of compromised systems, since you either didn't or else you assigned markedly low cost in order to be able to claim that the user convenience won out in this case.

You make various claims about what I've said, yet when I do the converse you demand that I back it up with a quote, even though I'm remarking on the notable _absence_ of a concern. I feel like the flaw is in your logic, not mine.

> Either it was obvious or it took more than a few seconds thought. You can't have it both ways.

Certainly something can be obvious only after a few seconds thought (or more). In any case, I never said anything about it being immediately obvious - I said that I was surprised people took a contrary position. You again twist my words to serve your purpose, which at this stage, seems to be just to insult me as much as possible.

I have not just "considered the benefits of one alternative and the costs of another". Your belief that I did so is misconstrued. You complain that I "whine about semantics" and yet (when not trying to insult me) you do little else than argue semantics. You continuously make personal insinuations and imply arrogance, preferring personal attacks over reasoned discussion. You use misplaced metaphors and flawed reasoning, relying on things that I haven't actually said, but you demand that I back up my re-statement of your own position with "a quote". There are examples of all this right through the discussion above, and you somehow think that a couple more flawed arguments with some personal insults thrown in wins the whole argument... I'm almost lost for words. And I _am_ done, absolutely this time, I'll grit my teeth and let whatever idiocy you want to spout afterwards stand on its own merit, or lack thereof.

**Comment by samnazarko** [comment]
I am locking this, as I feel nothing useful can be further added, and @davmac314 has already taken to asserting his opinion on other mediums [1]. I am far more interested in a serious discussion around the matter, as well as exploring how to make the system more secure in its default, shipped configuration, than to engage in baseless ad hominem. I will reiterate for a final time that we *have* made changes to the way SSH is enabled and presented to the user, and will make *further* changes. But further discussion here today seems to have been centered heavily about getting the last word in.

There should also be an additional dialog displayed if a user connects their device directly to the Internet without NAT to advise them that this configuration is not recommended. 

I am interested in improving OSMC in all areas, not arguing over functionality vs security. 

Thanks to those who have expressed concerns regarding our original implementation of SSH and have in turn, helped make OSMC more secure.

 -- S

[1] https://twitter.com/davmac314/status/709747439469727745
